### DISCUSSION TITLE
improve fifo performance?

### DISCUSSION THREAD

**Discussion by myklemykle** [earlephilhower/arduino-pico#1411] [Q&A]
Hi, 

I'm profiling the main loop in a multicore pico program that sends a single 32-bit value from core0 to core1 on every loop.

The entire loop takes about 300 usec to finish, on average.  My profiling suggests that a single call to rp2040.fifo.push_nb(), to send 32 bits from core0 to core1, is taking about 30 usec to complete, on a chip that's running at 133mhz.  That comes to about 4000 clock cycles to complete the non-blocking function.  That's a tenth of the loop, for one line of code.  Is that the expected performance of this fifo implementation? Anything I can to to make it faster?

Thanks in advance,
-mykle-

p.s. In general I am really loving the arduino-pico environment!  Works great, with great documentation.

**Comment by earlephilhower** [comment]
The FIFO now is built for correctness, not for speed, sorry.  Because we need to use the HW FIFO to cause an IRQ on the other core for freeze-CPU operations (i.e. while doing flash erase/write) it doesn't use the HW FIFO for `rp2040.push()`/etc.

It uses the Pico-SDK's multi-core, multi-thread safe queues: https://github.com/earlephilhower/arduino-pico/blob/f33dfd2313de853af8f50650fd2f6dabf788130d/cores/rp2040/RP2040Support.h#L70-L84

If you have a simple 1-producer/1-consumer setup it should be possible for you to implement a simple, lockless FIFO (as done in the SerialUART code) in your own code, and not use the built-in heavy duty queueing functions.

**Reply by myklemykle** [reply]
Fair enough. That's probably the right approach for this situation -- one producer, one consumer.
